src/util/side_metadata/sanity.rs

```diff
@@ -0,0 +1,738 @@
+                if v.is_err() {
+                    return v;
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+impl SideMetadataSanity {
+    /// Creates a new SideMetadataSanity instance.
+    pub fn new() -> SideMetadataSanity {
+        SideMetadataSanity {
+            specs_sanity_map: HashMap::new(),
+        }
+    }
+    /// Returns all global or policy-specific specs based-on the input argument.
+    ///
+    /// Returns a vector of globals if `global` is true and a vector of locals otherwise.
+    ///
+    /// Arguments:
+    /// * `global`: a boolean to show whether global (`true`) or policy-specific (`false`) specs are required.
+    ///
+    fn get_all_specs(&self, global: bool) -> Vec<SideMetadataSpec> {
+        let mut specs = vec![];
+        for (k, v) in self.specs_sanity_map.iter() {
+            if !(global ^ (*k == GLOBAL_META_NAME)) {
+                specs.append(&mut (*v).clone());
+            }
+        }
+
+        specs
+    }
+
+    /// Verifies that all local side metadata specs:
+    /// 1 - are not too big,
+    /// 2 - do not overlap.
+    ///
+    /// Returns `Ok(())` if no issue is detected, or `Err` otherwise.
+    ///
+    fn verify_local_specs(&self) -> Result<()> {
+        let local_specs = self.get_all_specs(false);
+
+        let v = verify_local_specs_size(&local_specs);
+        if v.is_err() {
+            return v;
+        }
+
+        for spec_1 in &local_specs {
+            for spec_2 in &local_specs {
+                if spec_1 != spec_2 {
+                    #[cfg(target_pointer_width = "64")]
+                    let v = verify_no_overlap_contiguous(spec_1, spec_2);
+                    #[cfg(target_pointer_width = "32")]
+                    let v = verify_no_overlap_chunked(spec_1, spec_2);
+                    if v.is_err() {
+                        return v;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Ensure that a metadata context does not have any issues for a space.
+    /// Panics with a suitable message if any issue is detected.
+    /// It also initialises the sanity maps which will then be used if the `extreme_assertions` feature is active.
+    /// Internally this calls verify_metadata_context().
+    ///
+    /// This function is called once per space but may be called multiple times per policy.
+    ///
+    /// Arguments:
+    /// * `space`: The space to check.
+    pub fn verify_space<VM: VMBinding, S: Space<VM>>(&mut self, space: &S) {
+        self.verify_metadata_context(
+            std::any::type_name::<S>(),
+            space.get_side_metadata_context(),
+        )
+    }
+
+    /// Ensure that metadata context used in spaces in CommonPlan (including BasePlan) also comply with our assumption.
+    /// If a plan uses CommonPlan, it needs to invoke this method .
+    pub fn verify_common_spaces<VM: VMBinding>(&mut self, common: &CommonPlan<VM>) {
```

Comment:
I think we'd better move `verify_common_spaces` and `verify_base_spaces` into the relevant `Space::new` methods.
Currently we're leaking the internal structure of common and base spaces

Reply:
Yeah. What you said sounds fine. For the current code, all the code is still within the crate - I am not worrying too much about encapsulation/leaking (if common/base spaces change, we can easily know and change the code accordingly).